libkbfs/backpressure_disk_limiter.go

```diff
@@ -239,6 +240,310 @@ func (bt *backpressureTracker) getStatus() backpressureTrackerStatus {
+// multiplied with the maximum delay to get the backpressure delay to
+// apply.
+func (qbt quotaBackpressureTracker) delayScale() float64 {
+	usedFrac := qbt.usedFrac()
+
+	// We want the delay to be 0 if usedFrac <= m and the max
+	// delay if usedFrac >= M, so linearly interpolate the delay
+	// scale.
+	//
+	// TODO: Consider a scheme which applies backpressure
+	// differently if R/Q > m, but (U+R)/Q <= 1. In that case, we
+	// don't want to "permanently" apply backpressure before the
+	// quota is actually hit. In that case, the backpressure
+	// should probably be proportional to U/(Q-R) until (U+R)/Q =
+	// 1, at which point it should smoothly transition to a
+	// formula that is proportional to (U+R)/Q.
+	m := qbt.minThreshold
+	M := qbt.maxThreshold
+	return math.Min(1.0, math.Max(0.0, (usedFrac-m)/(M-m)))
+}
+
+func (qbt *quotaBackpressureTracker) onJournalEnable(unflushedBytes int64) {
+	qbt.unflushedBytes += unflushedBytes
+}
+
+func (qbt *quotaBackpressureTracker) onJournalDisable(unflushedBytes int64) {
+	qbt.unflushedBytes -= unflushedBytes
+}
+
+func (qbt *quotaBackpressureTracker) updateRemote(
+	remoteUsedBytes, quotaBytes int64) {
+	qbt.remoteUsedBytes = remoteUsedBytes
+	qbt.quotaBytes = quotaBytes
+}
+
+func (qbt *quotaBackpressureTracker) afterBlockPut(
+	blockBytes int64, putData bool) {
+	if putData {
+		qbt.unflushedBytes += blockBytes
+	}
+}
+
+func (qbt *quotaBackpressureTracker) onBlocksFlush(blockBytes int64) {
+	qbt.unflushedBytes -= blockBytes
+}
+
+type quotaBackpressureTrackerStatus struct {
+	// Derived numbers.
+	UsedFrac   float64
+	DelayScale float64
+
+	// Constants.
+	MinThreshold float64
+	MaxThreshold float64
+
+	// Raw numbers.
+	UnflushedBytes  int64
+	RemoteUsedBytes int64
+	QuotaBytes      int64
+}
+
+func (qbt *quotaBackpressureTracker) getStatus() quotaBackpressureTrackerStatus {
+	return quotaBackpressureTrackerStatus{
+		UsedFrac:   qbt.usedFrac(),
+		DelayScale: qbt.delayScale(),
+
+		MinThreshold: qbt.minThreshold,
+		MaxThreshold: qbt.maxThreshold,
+
+		UnflushedBytes:  qbt.unflushedBytes,
+		RemoteUsedBytes: qbt.remoteUsedBytes,
+		QuotaBytes:      qbt.quotaBytes,
+	}
+}
+
+// journalTracker aggregates all the journal trackers.
+type journalTracker struct {
+	byte, file *backpressureTracker
+	quota      *quotaBackpressureTracker
+}
+
+func newJournalTracker(
+	minThreshold, maxThreshold, quotaMinThreshold, quotaMaxThreshold, journalFrac float64,
+	byteLimit, fileLimit, freeBytes, freeFiles int64) (
+	journalTracker, error) {
+	// byteLimit and fileLimit must be scaled by the proportion of
+	// the limit that the journal should consume.
+	journalByteLimit := int64((float64(byteLimit) * journalFrac) + 0.5)
+	byteTracker, err := newBackpressureTracker(
+		minThreshold, maxThreshold, journalFrac, journalByteLimit,
+		freeBytes)
+	if err != nil {
+		return journalTracker{}, err
+	}
+	// the fileLimit is only used here, but in the interest of
```

Comment:
Oops, the comment doesn't make sense anymore with the refactor. Updated it to say that the file limit is only used by the journal (i.e. not the cache).